### PR TITLE
adding cors handler

### PR DESCRIPTION
**Pull request by akoserwal**
### PR Template:

* [x] Are the agreed upon acceptance criteria fulfilled?

* [ ] Was the 4-eye-principle applied? (async PR review, pairing, ensembling)

* [ ] Do your changes have passing automated tests and sufficient observability?

* [ ] Are the work steps you introduced repeatable by others, either through automation or documentation?
  * [ ] If automation is possible but not done due to other constraints, a ticket to the tech debt sprint is added
  * [ ] An SOP (Standard Operating Procedure) was created

* [ ] The Changes were automatically built, tested, and  - if needed, behind a feature flag - deployed to our production environment. (**Please check this when the new deployment is done and you could verify it.**)

* [ ] Are the agreed upon coding/architectural practices applied?

* [x] Are security needs fullfilled? (e.g. no internal URL)

* x] Is the corresponding Ticket in the right state? (should be on "review" now, put to done when this change made it to production)

* [ ] For changes to the public API / code dependencies: Was the whole team (or a sufficient amount of ppl) able to review?

### CODE REVIEW

File: api/http/Server.go [hunk 1]
```diff
@@ -38,7 +46,7 @@ func (s *Server) Serve(wait *sync.WaitGroup) error {
 			err = http.ListenAndServeTLS(
 				":"+s.ServerConfig.HTTPSPort,
 				s.ServerConfig.TLSConfig.CertPath, //TODO: Needs sanity checking.
-				s.ServerConfig.TLSConfig.KeyPath, mux)
+				s.ServerConfig.TLSConfig.KeyPath, handler)
```
Comment (DGuhr) [inline api/http/Server.go:49]: I'd love to see some tests.  

Also: Was it intentional to add it only when SSL is set up and not [here](https://github.com/RedHatInsights/authz/pull/90/files#diff-db6116b4f5c82f4b81e77379964a2c6324130b00acc1ab9e1deacfa78d9b9a41R58) also? If so, I am not sure it is the best idea. A different approach could be to define the AllowedOrigins ( + perhaps Headers/Methods) in a configuration and load this for both.

File: api/http/Server.go [hunk 2]
```diff
@@ -30,6 +31,12 @@ func (s *Server) Serve(wait *sync.WaitGroup) error {
 		return err
 	}
 
+	handler := cors.New(cors.Options{
+		AllowedMethods: []string{"GET", "POST", "PATCH", "PUT", "DELETE"},
+		AllowedHeaders: []string{"Authorization", "Content-Type"},
+		MaxAge:         300,
+	}).Handler(mux)
```
Comment (DGuhr) [inline api/http/Server.go:38]: I think we also need to set `AllowedCredentials: true` to allow for  user credentials like cookies, HTTP authentication or client side SSL certificates, but not sure

File: api/http/Server.go [hunk 3]
```diff
@@ -30,6 +31,13 @@ func (s *Server) Serve(wait *sync.WaitGroup) error {
 		return err
 	}
 
+	handler := cors.New(cors.Options{
+		AllowedMethods:   []string{"GET", "POST", "PATCH", "PUT", "DELETE"},
+		AllowedHeaders:   []string{"Authorization", "Content-Type"},
```
Comment (DGuhr) [inline api/http/Server.go:36]: I'D also add common headers such as `Accept, Content-Type, Content-Length, Accept-Encoding, Authorization, ResponseType`

ref: e.g. https://fale.io/blog/2021/07/28/cors-headers-with-grpc-gateway

**Comment by Rajagopalan-Ranganathan** [comment]
I see the comments have been addressed @akoserwal can you resolve them?

**Comment by akoserwal** [comment]
> Why don't you define `AllowedOrigins` in `cors.New()`?
> 
> e.g. from https://github.com/rs/cors:
> 
> ```
> c := cors.New(cors.Options{
>     AllowedOrigins: []string{"http://foo.com", "http://foo.com:8080"},
>     AllowCredentials: true,
>     // Enable Debugging for testing, consider disabling in production
>     Debug: true,
> })
> ```
> 
> I assume that the default is to allow all origins, but I don't see why we would do that (although keeping the consoledot origin domain up to date/accurate is a job in itself).

For now, to keep it simple and easy to work with service. Skipping the allowed origin config. I'll create a follow jira to configure allow origin with flag/config file for later.

**Comment by merlante** [comment]
Most/all of the tests in `Server_test.go` currently use the mux created with `createMultiplexer()` in Server.go (from `runRequestWithServer()`). Can we bring the cors stuff into `createMultiplexer()`, so that we are still testing on the "final" mux?

**Comment by akoserwal** [comment]
> Most/all of the tests in `Server_test.go` currently use the mux created with `createMultiplexer()` in Server.go (from `runRequestWithServer()`). Can we bring the cors stuff into `createMultiplexer()`, so that we are still testing on the "final" mux?

Most of the test call `runRequest` which calls runRequestWithServer. Cors config is available for all the tests.
We cannot bring cors handler into createMultiplexer() unless change the return type http.Handler.

File: api/http/Server.go [hunk 4]
```diff
@@ -30,6 +31,13 @@ func (s *Server) Serve(wait *sync.WaitGroup) error {
 		return err
 	}
 
+	handler := cors.New(cors.Options{
+		AllowedOrigins: []string{},
```
[the remark this answers is not in the document]
Comment (DGuhr) [reply api/http/Server.go]: Guess this was accidentally resolved? (@merlante  had more or less the same question)

Looking at the libraries docs at https://github.com/rs/cors - without trying it out - I think either we need to add a `*` here, or  use `cors.Default()`, right?

btw: All in to use `*` in first iteration and use a configurable slice via flag/config later on.

File: api/http/Server.go (hunk 2 above)
Comment (DGuhr) [reply api/http/Server.go:39]: resolving this, as AllowCredentials seems to have been added.

File: api/http/Server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (DGuhr) [reply api/http/Server.go]: Looking at the code again: Did you remove the AllowedOrigins completely? What would the behaviour be in that case?

[the remark this answers is not in the document]
Comment (akoserwal) [reply api/http/Server.go]: Default value is ["*"], if we don't configure it

File: api/http/Server.go [hunk 5]
```diff
@@ -30,6 +31,13 @@ func (s *Server) Serve(wait *sync.WaitGroup) error {
 		return err
 	}
 
+	handler := cors.New(cors.Options{
+		AllowedMethods:   []string{"GET", "POST", "PATCH", "PUT", "DELETE"},
```
Comment (DGuhr) [inline api/http/Server.go:35]: given that we're working with a js frontend, I guess we'd also need the `OPTIONS` Method covered. Quote from thread: `... It’s the server that needs to respond to the OPTION with the Allow-origin set with something that matches the caller’s host`

Comment (DGuhr) [reply api/http/Server.go:35]: Let's just try it out without, and add it if needed

**Pull request closed without merging**